### PR TITLE
Add UMA execution to SafeSnap

### LINKED ISSUE #2814: Integrate UMA optimistic governance on SafeSnap

### PR DESCRIPTION
**Pull request by pemulis**
Fixes #2814 

**Changes in this PR:**
This pull request adds UMA execution as an option in SafeSnap. Our design segregates the UMA elements as much as possible from the existing Snapshot code, to make review easier and prevent the introduction of unexpected bugs.

In effect, the code should detect when the user has specified an UMA module address in their configuration, and detect a value tied to the UMA Optimistic Governor contract's `rules`. If the code sees that the user has an UMA module configured, they will be directed to a distinct `HandleOutcomeUma` view that uses the `umaModule` file.

There are also various configuration and other changes in the codebase, with a light touch, trying to leave Reality execution and everything else alone.

**How to test and review this PR?**
We've been testing functionality in a Vercel preview in our repo.
https://snapshot-uma.vercel.app/

The Outcome Test Snapshot Space is tied to a Safe with an UMA governance module and an UMA module address set in the configuration.
https://snapshot-uma.vercel.app/#/outcome.eth

I believe you can join the Space to create new proposals and test the end-to-end flow, but let me know if you run into any issues. The Space is set up on Goerli with bonds in Goerli ETH.

Edit: Looks like you have your own Vercel deployment flow but I don't have authorization to trigger it on this PR.

You should be able to successfully:

1. Create a proposal
2. Add transactions to the proposal
3. Vote on the proposal
4. Approve a bond
5. Request execution of a proposal (after the voting period)
6. Dispute execution through the testnet oracle UI (https://testnet.oracle.umaproject.org/)
7. Delete disputed proposals through the Snapshot UI
8. Propose again
9. Execute proposals after the challenge period if it is not disputed
10. See when a proposal has been completely executed
11. Create a different or duplicate proposal, and run through the same flow

If desired, we can also help you set up a new Safe, UMA module, and Snapshot Space to test the whole set-up from scratch. Until we have this merged, and a parallel PR merged with the Zodiac repo, the UMA module contracts will need to be deployed manually rather than through the Zodiac app.

In the end state, the user set-up flow will look like:
`Create Safe -> Create Snapshot Space -> Deploy UMA module through Zodiac app`

In other words, the user should be able to create an "optimistic DAO" without writing any code, through simple point-and-click interfaces. After that, users can handle governance actions through their familiar Snapshot interface.

### CODE REVIEW

**Comment by mktcode** [comment]
This deleted proposal: https://snapshot-uma.vercel.app/#/outcome.eth/proposal/0x2e5ba3fb5461664995f76fc5b3e87a113826a8fa98cf1a3ad674604676999394

I'm trying to re-propose it, payed the bond but when clicking on "Request execution" and get this error:
image

**Comment by pemulis** [comment]
> This deleted proposal: https://snapshot-uma.vercel.app/#/outcome.eth/proposal/0x2e5ba3fb5461664995f76fc5b3e87a113826a8fa98cf1a3ad674604676999394
> 
> I'm trying to re-propose it, payed the bond but when clicking on "Request execution" and get this error: image
> 
> Also old proposals break if they don't have UMA configured: http://localhost:8080/#/gnosis.eth/proposal/0x6a581c6d10bec9e58a9f0e2cfa4ca83ceddbcec789862cedc432173e54042b80 image [UMAprotocol#2](https://github.com/UMAprotocol/snapshot-execution/pull/2)
> 
> And lastly: [UMAprotocol#1](https://github.com/UMAprotocol/snapshot-execution/pull/1)

Thanks for your comments! I merged both of your PRs. The old proposal was probably broken as we adjusted the UMA module (Optimistic Governor) contract during testing. It might be referencing an outdated contract address, or something along those lines.

When testing this pull request to the main Snapshot repo, you should create new proposals from scratch and try each flow. The revised code handles rejected proposals through the same flow as disputed proposals.

**Comment by mktcode** [comment]
> The old proposal was probably broken as we adjusted the UMA module (Optimistic Governor) contract during testing.

Out if curiosity... What did change? Is it on some branch? [Here](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol) I don't see recent changes.

**Comment by pemulis** [comment]
> > The old proposal was probably broken as we adjusted the UMA module (Optimistic Governor) contract during testing.
> 
> Out of curiosity... What did change? Is it on some branch? [Here](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/gnosis-zodiac/implementation/OptimisticGovernor.sol) I don't see recent changes.

We just added a `ProposalExecuted` event to the contract that lets us distinguish between different proposals that have identical transaction bundles. We also pass the unique IPFS CID generated by Snapshot as the proposal `explanation`. 

https://github.com/UMAprotocol/protocol/pull/4374

By the way, are you able to authorize the Vercel preview deployment? It would be good to test any changes that come in on this PR.

**Comment by samuveth** [comment]
Authorized the deployment for you

**Comment by mktcode** [comment]
Went through it again. Couldn't find any more serious issues. The `ensureRightNetwork` function is duplicated and the `approveBond` and `getModuleDetailsUma` should be more strict about their parameters. Aside from such things... looks good.

Approving this PR and inviting you to share my thoughts with you based on the refactored code.

**Review by samuveth — APPROVED** [review]
utACK

**Comment by zzuziak** [comment]
@bonustrack could you have a look and merge if ready? :pray:

File: src/plugins/safeSnap/index.ts [hunk 1]
```diff
@@ -130,9 +137,106 @@ export default class Plugin {
...
-    return getModuleDetails(provider, network, moduleAddress);
+    return getModuleDetailsReality(provider, network, moduleAddress);
+  }
+
+  async validateUmaModule(network: string, umaAddress: string) {
+    if (!isAddress(umaAddress)) return 'reality';
+    
+    const provider: StaticJsonRpcProvider = getProvider(network);
+    const moduleContract = new Contract(
+      moduleAddress,
```
Comment (bonustrack) [inline src/plugins/safeSnap/index.ts:150]: The var `moduleAddress` does not exists here, I assume you mean `umaAddress`
```suggestion
      umaAddress,
```

**Comment by bonustrack** [comment]
I've tested the flow locally and it seem to work well. I've been able to create a proposal with execution, request execution, and execute txs.

The proposal is here:
http://localhost:8080/#/outcome.eth/proposal/0xfebd7c1d7b9102c01f43340d9c5712479f65fafe446cd532e219679b1b2b7e71
And execution:
https://goerli.etherscan.io/tx/0xec011de9f075a29837e41ebd2319a7f9d1e9d39a8e4ff116c58c1643dc85b2f0

One thing I've not been able to do is to make dispute, I can't connect my wallet when I go here: https://testnet.oracle.umaproject.org/request?transactionHash=0x9f1a7b2cf807aec95a128e1ebabae5ed9b96dc80081f9627c63d1d02917759a4&chainId=5&oracleType=OptimisticV2&eventIndex=0
I tried with a wallet on Goerli and mainnet but it doesn't login me for some reason (there is no error in the console).

Also any reason to have a button "Click to confirm the proposal passed"? Is it to avoid someone request execution when proposal did not pass?

**Comment by pemulis** [comment]
> I've tested the flow locally and it seem to work well. I've been able to create a proposal with execution, request execution, and execute txs.
> 
> The proposal is here: http://localhost:8080/#/outcome.eth/proposal/0xfebd7c1d7b9102c01f43340d9c5712479f65fafe446cd532e219679b1b2b7e71 And execution: https://goerli.etherscan.io/tx/0xec011de9f075a29837e41ebd2319a7f9d1e9d39a8e4ff116c58c1643dc85b2f0
> 
> One thing I've not been able to do is to make dispute, I can't connect my wallet when I go here: https://testnet.oracle.umaproject.org/request?transactionHash=0x9f1a7b2cf807aec95a128e1ebabae5ed9b96dc80081f9627c63d1d02917759a4&chainId=5&oracleType=OptimisticV2&eventIndex=0 I tried with a wallet on Goerli and mainnet but it doesn't login me for some reason (there is no error in the console).
> 
> Also any reason to have a button "Click to confirm the proposal passed"? Is it to avoid someone request execution when proposal did not pass?

We're looking into the OO UI issue to see what's going on, but that is unrelated to this PR. The button to confirm the proposal passed is just to avoid someone proposing something that didn't actually get approved on Snapshot, since (afaik) there is nothing in the Snapshot state that tells us if the off-chain vote approved or disapproved. We were on the fence about including that extra step, but we want to keep people from losing money on a proposal bond due to confusion.

**Comment by pemulis** [comment]
If the flow looks good to you, though, I think we're okay to merge and can look into the Optimistic Oracle UI issue separately. Or, if you want to test the post-dispute flows, you can let us know when you create a new proposal and we can dispute it manually (if the OO UI is still having trouble on testnet).

**Comment by mktcode** [comment]
> since (afaik) there is nothing in the Snapshot state that tells us if the off-chain vote approved or disapproved.

*(Thought on this for future improvements:)* The criteria should be defined together with the execution data. A batch of transactions can have a "criteria" field specifying human-readable execution criteria, considering the fact that ultimate judgement, in case of dispute, will also be human. It's up to the author to make sure the criteria is not ambiguous. The criteria can be displayed alongside this button to confirm the proposal.

**Review by bonustrack — APPROVED** [review]
